Add FDBWorkloadContext::delay

Wonshtrum opened the pull request:
Implementation proposal for issue #12343.
Continuation of #11288.

I added a `delay` method on the `FDBWorkloadContext` of the C simulation API.
It returns a `FDBFuture` pointer and can be treated like any other futures returned by the `fdb_c` API.
Two problems still need to be addressed:
- should we update the C++ simulation API (updating the `CppWorkload.h` to also include `delay`)?
- how to make the C simulation API more resilient to API changes?

The first point is quite straightforward, I can provide the C++ implementation easily, but what about the Java bindings? Should we update them as well?

For the second point, the problem is that the full API is a list of function pointers the fdbserver sends to the workload as a struct on the stack:
https://github.com/apple/foundationdb/blob/a2940a7e891f8f2a63b09ada5a2f6ab08e49a6a7/bindings/c/test/workloads/CWorkload.c#L87
https://github.com/apple/foundationdb/blob/a2940a7e891f8f2a63b09ada5a2f6ab08e49a6a7/bindings/c/foundationdb/CWorkload.h#L85-L101
Adding a new binding, like `delay`, is a breaking change since it changes the size of the struct. A simple way to fix this, I think, would be to pass a pointer to the `FDBWorkloadContext` instead of the struct directly. If we only add entries in this struct, never removing or swapping fields, older workloads can still run as they access the bindings by their offset in the context. Newer workloads can use newer bindings located at higher offsets in the struct

This was an oversight on my part. I hope we can merge both the `delay` addition and the "by pointer" API in 7.4 before the end of the release candidate phase, so the API is fixed before it’s officially released.

spraza commented:
> Hi @spraza, since you merged the first version of the C API, I would really appreciate it if you could take a look at this.

I'll review this change next week but keep in mind that this specific PR is for main branch. Once the release branches are cut, we generally only allow critical changes like bug fixes. Once I review the PR, we can determine if we can backport the change to 7.4 branch.

Wonshtrum commented:
I'm very sorry, I changed the API again...
If you remember, on the first PR I wasn't sure what kind of layout to use for the API, and we went with this "object-oriented" design in the end. Here, I take it a step further, collecting the function pointers of each object in virtual tables, very similar to how C++ does dynamic dispatch, or even closer to Rust "fat pointers" (`&dyn trait`), which are a pointer to data and a pointer to a table together.
Here, "objects" are a tuple `struct Type { OpaqueType* inner; Type_VT vt; }`.
If the API changes, we can add new methods to a type's virtual table without breaking older workloads.
So a workload compiled for 7.4 would also run on a 7.5 fdbserver, even if new methods (like `delay`) were added.

I also updated the Rust workload, since it revealed issues last time. Here it works like a charm. I did not implement the `delay` though, since I would need the Rust glue to work with `FDBFuture`. It was already implemented successfully in foundationdb-rs I don't think bloating this example makes sense here.

spraza commented:
@Wonshtrum I am finally getting to this. Thanks for your patience. Initial review looks good. But I want to test this on my end before accepting. Can you mention in the PR description how you're testing this?

spraza commented on bindings/c/foundationdb/CWorkload.h:
```diff
@@ -57,60 +67,67 @@ typedef struct FDBMetric {
+		int (*clientCount)(OpaqueWorkloadContext* inner);
+		// Return a random seed (same each time and for all clients).
+		int64_t (*sharedRandomNumber)(OpaqueWorkloadContext* inner);
+		// Return a future that will be resolved after a given time.
```
nit: s / resolved / ready /

spraza commented on bindings/c/foundationdb/CppWorkload.h:
```diff
@@ -64,6 +64,7 @@ class FDBWorkloadContext : public FDBLogger {
 	virtual int clientId() const = 0;
 	virtual int clientCount() const = 0;
 	virtual int64_t sharedRandomNumber() const = 0;
+	virtual FDBFuture* delay(double seconds) const = 0;
```
Fyi, flow delay accepts priority as well: https://github.com/apple/foundationdb/blob/1d03904ea3a2a47222445c0f2fdb89c95ed2631a/flow/include/flow/flow.h#L1570. 

I think it's fine to stick with default for now.

spraza replied:
It's not used that much, especially in the context of test workloads. I would leave this code as is to reduce the complexity in this PR. We can document that default priority is used as a comment in the function declaration. Later, if we've a use-case, we can always add priority and at that point, think about the details of how to expose it.

spraza reviewed (approved):
Looks good overall, 1 fyi and 1 minor nit. Agree this is very useful. I think it's worthwhile to backport to 7.4 for these reasons:
- hard to live without `delay`, so this is not a "feature" in the "new functionality" sense. I can't imagine writing workloads without having `delay` in my toolset.
- risk for breakage is really low given this is a new capability. Also, this does not affect native/core fdb simulation and definitely does not affect anything prod related
- we're still in pre-release 7.4 state

We can make an exception for this case. Once merged, I can backport this PR to 7.4 for you.

spraza reviewed (changes requested):
Looks like a ctest is failing on arm: https://github.com/apple/foundationdb/pull/12357#issuecomment-3304614661

spraza commented:
> I'm really not sure of what's wrong here. To sum it up:
> 
> * `foundationdb-pr-clang on Linux RHEL 9`: failed 1 time for timeout on s3_backup_tests, then succeeded 2 times
> * `foundationdb-pr-clang-arm on Linux CentOS 7`: succeeded, then failed on fdb_c_api_test_CApiCorrectnessCallbacksOnExtThr (but all logs are severity=10), then succeeded again
> * `foundationdb-pr-macos on macOS Ventura 13.x` and `foundationdb-pr-macos-m1 on macOS Ventura 13.x` failed to build 2 times

First and last ones are not related to your change, most likely. Middle one I debugged and it looks like the workload was stuck somewhere and couldn't complete. But I haven't seen the issue on x86_64 (I repro'ed) and even on arm, it has been passing since then. 

Overall, I don't think these are blockers. We can merge this PR.

Wonshtrum commented:
I replaced `resolved` by `ready` and added a note about `TaskPriority::DefaultPriority`.
This looks good to me. I hope we won't discover a new flaw in this API before long.
Thank you again for the review. I hope the backport will go smoothly. We will extensively use this API on our side.
